[Site Isolation] Properly create scrolling tree for web process with multiple root frames

nmoucht opened the pull request:
#### 48aa61d62068efcceb7aebce99d86384e5d2f0f0
<pre>
[Site Isolation] Properly create scrolling tree for web process with multiple root frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=272721">https://bugs.webkit.org/show_bug.cgi?id=272721</a>
<a href="https://rdar.apple.com/126001790">rdar://126001790</a>

Reviewed by Simon Fraser.

Adjust AsyncScrollingCoordinator to hold a scrolling state tree per root frame.
Also thanks to Alex who helped take this pr over the finishing line.

* LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree.html:
* LayoutTests/http/tests/site-isolation/scrolling/multiple-root-frames-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/multiple-root-frames.html: Copied from LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree.html.
* LayoutTests/http/tests/site-isolation/scrolling/remove-root-frame-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/remove-root-frame.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/SmallMap.h: Added.
(WTF::SmallMap::ensure):
(WTF::SmallMap::remove):
(WTF::SmallMap::get const):
(WTF::SmallMap::forEach const):
(WTF::SmallMap::size const):
(WTF::SmallMap::rawStorage const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setBackForwardCacheState):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::~LocalFrame):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::rootFrameID const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::AsyncScrollingCoordinator):
(WebCore::AsyncScrollingCoordinator::stateNodeForNodeID const):
(WebCore::AsyncScrollingCoordinator::stateNodeForScrollableArea const):
(WebCore::AsyncScrollingCoordinator::ensureScrollingStateTreeForRootFrameID):
(WebCore::AsyncScrollingCoordinator::existingScrollingStateTreeForRootFrameID const):
(WebCore::AsyncScrollingCoordinator::rootFrameWasRemoved):
(WebCore::AsyncScrollingCoordinator::stateTreeForNodeID const):
(WebCore::AsyncScrollingCoordinator::willCommitTree):
(WebCore::AsyncScrollingCoordinator::updateEventTrackingRegions):
(WebCore::AsyncScrollingCoordinator::frameViewLayoutUpdated):
(WebCore::AsyncScrollingCoordinator::frameViewEventTrackingRegionsChanged):
(WebCore::AsyncScrollingCoordinator::frameViewForScrollingNode const):
(WebCore::AsyncScrollingCoordinator::createNode):
(WebCore::AsyncScrollingCoordinator::insertNode):
(WebCore::AsyncScrollingCoordinator::unparentNode):
(WebCore::AsyncScrollingCoordinator::unparentChildrenAndDestroyNode):
(WebCore::AsyncScrollingCoordinator::detachAndDestroySubtree):
(WebCore::AsyncScrollingCoordinator::clearAllNodes):
(WebCore::AsyncScrollingCoordinator::reconcileViewportConstrainedLayerPositions):
(WebCore::AsyncScrollingCoordinator::ensureRootStateNodeForFrameView):
(WebCore::AsyncScrollingCoordinator::hasSubscrollers const):
(WebCore::AsyncScrollingCoordinator::scrollingStateTreeAsText const):
(WebCore::AsyncScrollingCoordinator::commitTreeStateForRootFrameID):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
(WebCore::AsyncScrollingCoordinator::scrollingStateTrees const):
(WebCore::AsyncScrollingCoordinator::scrollingStateTree): Deleted.
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::rootFrameWasRemoved):
(WebCore::ScrollingCoordinator::createNode):
(WebCore::ScrollingCoordinator::insertNode):
(WebCore::ScrollingCoordinator::clearAllNodes):
(WebCore::ScrollingCoordinator::hasSubscrollers const):
(WebCore::ScrollingCoordinator::willCommitTree):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::removeNode):
(WebCore::ScrollingTree::commitTreeStateInternal):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.cpp:
(WebCore::ScrollingTreeFrameHostingNode::removeHostedChild):
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.h:
* Source/WebCore/page/scrolling/ThreadedScrollingCoordinator.cpp:
(WebCore::ThreadedScrollingCoordinator::commitTreeStateIfNeeded):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::rootFrameID const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::maintainsEventRegion const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingLayers):
(WebCore::RenderLayerCompositor::attachWidgetContentLayersIfNecessary):
(WebCore::RenderLayerCompositor::registerScrollingNodeID):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::rootFrameID const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::rootFrameID const):
* Source/WebCore/rendering/RenderListBox.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::rootFrameID const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createScrollingNodeIfNecessary):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::buildTransaction):
* Tools/TestWebKitAPI/Tests/WTF/SmallSet.cpp:
(TestWebKitAPI::TEST(WTF_SmallMap, Basic)):

Canonical link: <a href="https://commits.webkit.org/278591@main">https://commits.webkit.org/278591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/269b787fa85a094c32b5ff0b827f376488c5c64c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51000 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3326 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54257 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1689 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53301 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36585 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1355 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41552 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53098 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27941 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43941 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22670 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25276 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1196 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9440 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44338 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47256 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1270 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55851 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50501 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26107 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1167 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48953 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27358 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44012 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48079 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28236 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/57976 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7409 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27089 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11935 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -242,6 +245,16 @@ bool LocalFrame::isRootFrame() const
+    if (auto* localMainFrame = dynamicDowncast<LocalFrame>(mainFrame()))
+        return *localMainFrame;
+    auto* iterator = this;
+    while (auto* localParent = dynamicDowncast<LocalFrame>(iterator->tree().parent()))
```
Can't we have `Page` or someone hold onto the root frame, rather than having to climb the tree every time?

achristensen07 replied:
The Page does have m_rootFrames, but we don't know which root frame is this frame's root frame.  For Frame::mainFrame we store a pointer m_mainFrame on each Frame.  We could do the same for this if it were common enough to be worth the memory use.

nmoucht replied:
This is fixed after https://github.com/WebKit/WebKit/pull/27823.

smfr commented on Source/WebCore/page/LocalFrame.h:
```diff
@@ -170,6 +170,7 @@ class LocalFrame final : public Frame {
     void resetScript();
 
     WEBCORE_EXPORT bool isRootFrame() const final;
+    WEBCORE_EXPORT LocalFrame& rootFrame();
```
This would be clearer as `rootmostLocalFrame()`

achristensen07 replied:
I chose rootFrame to be more similar to Frame::mainFrame.  All root frames are LocalFrames, so rootmostLocalFrame would be redundant naming.

smfr replied:
"All root frames are LocalFrames" — that may be clear to you, but it's not to me.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -107,6 +106,40 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
 }
 #endif
 
+ScrollingStateTree& AsyncScrollingCoordinator::ensureScrollingStateTreeForRootFrameID(FrameIdentifier identifier)
```
This needs to assert that `identifier` is a root frame identifier. Please also rename the parameter.

achristensen07 replied:
There is currently no way to assert that it is a root frame identifier, but we could easily add such a way by keeping extra bookkeeping but only if ASSERT_ENABLED is true.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -107,6 +106,40 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
+        return it->value.get();
+
+    ASSERT_WITH_MESSAGE(!page() || page()->settings().siteIsolationEnabled() || m_scrollingStateTrees.isEmpty(), "With site isolation off, we should only ever have one scoling state tree. Get the frame identifier from LocalFrame::rootFrame.");
+    auto addResult = m_scrollingStateTrees.set(identifier, makeUniqueRef<ScrollingStateTree>(this));
```
You can use the `it` you got before to do the insertion. `ensure()` can be a better way to write this.

nmoucht replied:
Added use of ensure.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -107,6 +106,40 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
+    return addResult.iterator->value.get();
+}
+
+void AsyncScrollingCoordinator::removeRootFrame(FrameIdentifier identifier)
```
Call this `rootFrameWasRemoved()` since this code does not actually remove the root frame.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -107,6 +106,40 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
+    return nullptr;
+}
+
+ScrollingStateTree* AsyncScrollingCoordinator::stateTreeForID(ScrollingNodeID nodeID) const
```
`stateTreeForNodeID()` to disambiguate from other kinds of IDs.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -107,6 +106,40 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
+    m_scrollingStateTrees.remove(identifier);
+}
+
+RefPtr<ScrollingStateNode> AsyncScrollingCoordinator::stateNodeForID(ScrollingNodeID nodeID) const
```
`stateNodeForNodeID()` to disambiguate from other kinds of IDs.

I'm not really a fan of having to do lookups in all the scrolling trees, since this gets called a lot. In many call sites we might already have the root frame identifier?

nmoucht replied:
I think it's preferable to do this versus having to go through frameViewForScrollingNode. For the call sites with the frame view we can maybe get away with getting the root frame id that way.

nmoucht replied:
After https://github.com/WebKit/WebKit/pull/27823 I added roofFrameID to ScrollableArea, so for cases where we have the ScrollableArea we can avoid looking through the scrolling trees.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -246,11 +282,11 @@ void AsyncScrollingCoordinator::frameViewRootLayerDidChange(LocalFrameView& fram
     // If the root layer does not have a ScrollingStateNode, then we should create one.
     ensureRootStateNodeForFrameView(frameView);
-    ASSERT(m_scrollingStateTree->stateNodeForID(frameView.scrollingNodeID()));
+    ASSERT(ensureScrollingStateTreeForRootFrameID(frameView.frame().rootFrame().frameID()).stateNodeForID(frameView.scrollingNodeID()));
```
It's bad to call an `ensure` function inside an assertion, because you've now got a debug-only behavior difference.

nmoucht replied:
Removed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -246,11 +282,11 @@ void AsyncScrollingCoordinator::frameViewRootLayerDidChange(LocalFrameView& fram
     ScrollingCoordinator::frameViewRootLayerDidChange(frameView);
 
-    RefPtr node = dynamicDowncast<ScrollingStateFrameScrollingNode>(m_scrollingStateTree->stateNodeForID(frameView.scrollingNodeID()));
+    RefPtr node = dynamicDowncast<ScrollingStateFrameScrollingNode>(ensureScrollingStateTreeForRootFrameID(frameView.frame().rootFrame().frameID()).stateNodeForID(frameView.scrollingNodeID()));
```
Avoid long call chains; pull the `ensureScrollingStateTreeForRootFrameID()` out onto a previous line.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -118,20 +151,21 @@ static inline void setStateScrollingNodeSnapOffsetsAsFloat(ScrollingStateScrolli
         return;
 
-    if (!m_scrollingStateTree->rootStateNode())
+    auto it = m_scrollingStateTrees.find(rootFrameID);
```
Maybe add `existingScrollingStateTreeForRootFrameID()` and use it here.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -499,12 +535,16 @@ void AsyncScrollingCoordinator::scheduleRenderingUpdate()
 {
     ASSERT(rootFrame.isRootFrame());
-    if (scrollingNodeID == m_scrollingStateTree->rootStateNode()->scrollingNodeID()) {
+    auto it = m_scrollingStateTrees.find(rootFrame.frameID());
```
Maybe add `existingScrollingStateTreeForRootFrameID()` and use it here.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -821,44 +861,48 @@ void AsyncScrollingCoordinator::scrollableAreaScrollbarLayerDidChange(Scrollable
 }
 
-ScrollingNodeID AsyncScrollingCoordinator::createNode(ScrollingNodeType nodeType, ScrollingNodeID newNodeID)
+ScrollingNodeID AsyncScrollingCoordinator::createNode(FrameIdentifier rootID, ScrollingNodeType nodeType, ScrollingNodeID newNodeID)
```
`rootID` -> `rootFrameID` (prefer variable names with the least amount of ambiguity).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -821,44 +861,48 @@ void AsyncScrollingCoordinator::scrollableAreaScrollbarLayerDidChange(Scrollable
 }
 
-ScrollingNodeID AsyncScrollingCoordinator::insertNode(ScrollingNodeType nodeType, ScrollingNodeID newNodeID, ScrollingNodeID parentID, size_t childIndex)
+ScrollingNodeID AsyncScrollingCoordinator::insertNode(FrameIdentifier rootID, ScrollingNodeType nodeType, ScrollingNodeID newNodeID, ScrollingNodeID parentID, size_t childIndex)
```
`rootFrameID`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -1074,7 +1118,7 @@ void AsyncScrollingCoordinator::setSynchronousScrollingReasons(ScrollingNodeID n
 OptionSet<SynchronousScrollingReason> AsyncScrollingCoordinator::synchronousScrollingReasons(ScrollingNodeID nodeID) const
 {
-    RefPtr node = dynamicDowncast<ScrollingStateScrollingNode>(m_scrollingStateTree->stateNodeForID(nodeID));
+    RefPtr node = dynamicDowncast<ScrollingStateScrollingNode>(stateNodeForID(nodeID));
```
This patch would be much smaller if you did all the `stateNodeForID()` changes first.

achristensen07 replied:
That's how it started but it wasn't testable without the rest.  Now that we know what it takes to get the rest, it could be fine to split it up.  No preference from me.

nmoucht replied:
Planning on splitting this up now.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -1093,7 +1137,11 @@ void AsyncScrollingCoordinator::windowScreenDidChange(PlatformDisplayID displayI
 
 bool AsyncScrollingCoordinator::hasSubscrollers() const
```
It seems that this can only be answered correctly for the main frame.

nmoucht replied:
Updated this to include root frame id.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
```diff
@@ -78,21 +78,23 @@ class AsyncScrollingCoordinator : public ScrollingCoordinator {
     void setScrollingTree(Ref<ScrollingTree>&& scrollingTree) { m_scrollingTree = WTFMove(scrollingTree); }
-
-    ScrollingStateTree* scrollingStateTree() { return m_scrollingStateTree.get(); }
+    HashMap<FrameIdentifier, UniqueRef<ScrollingStateTree>>& scrollingStateTrees() { return m_scrollingStateTrees; }
```
const function returning a const ref?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
```diff
@@ -78,21 +78,23 @@ class AsyncScrollingCoordinator : public ScrollingCoordinator {
     WEBCORE_EXPORT bool haveScrollingTree() const override;
 
-    WEBCORE_EXPORT void willCommitTree() override;
+    WEBCORE_EXPORT void willCommitTree(FrameIdentifier) override;
```
Is FrameIdentifier a rootFrameID?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
```diff
@@ -111,6 +113,7 @@ class AsyncScrollingCoordinator : public ScrollingCoordinator {
     WEBCORE_EXPORT void frameViewVisualViewportChanged(LocalFrameView&) override;
     WEBCORE_EXPORT void frameViewEventTrackingRegionsChanged(LocalFrameView&) override;
     WEBCORE_EXPORT void frameViewWillBeDetached(LocalFrameView&) override;
+    WEBCORE_EXPORT void removeRootFrame(FrameIdentifier) final;
```
Is FrameIdentifier a rootFrameID?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -345,17 +346,19 @@ bool ScrollingTree::commitTreeStateInternal(std::unique_ptr<ScrollingStateTree>&
     CommitTreeState commitState;
 
     if (hostingContextIdentifier) {
-        LOG_WITH_STREAM(Scrolling, stream << "ScrollingTree::commitTreeState - starting hosted tree commit for hosting context ID:  " << *hostingContextIdentifier);
```
Why remove this?

nmoucht replied:
Re-added.

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.cpp:
```diff
@@ -73,7 +73,7 @@ void ScrollingTreeFrameHostingNode::removeHostedChildren()
     auto hostedChildren = std::exchange(m_hostedChildren, { });
     for (auto& children : hostedChildren)
-        scrollingTree().removeNode(children->scrollingNodeID());
+        scrollingTree().removeNode(children->scrollingNodeID(), nullptr);
```
Just provide a default argument to `removeNode()`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.cpp:
```diff
@@ -83,6 +83,13 @@ void ScrollingTreeFrameHostingNode::willBeDestroyed()
+void ScrollingTreeFrameHostingNode::removeHostedChild(RefPtr<ScrollingTreeNode> node)
+{
+    m_hostedChildren.remove(node);
+    if (node)
```
Why not the if check for the line above?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -2700,7 +2700,7 @@ auto RenderLayerCompositor::attachWidgetContentLayersIfNecessary(RenderWidget& r
             if (auto pluginScrollingNodeID = renderEmbeddedObject->scrollingNodeID()) {
                 if (isVisible) {
-                    scrollingCoordinator->insertNode(ScrollingNodeType::PluginScrolling, pluginScrollingNodeID, pluginHostingNodeID, 0);
+                    scrollingCoordinator->insertNode(m_renderView.frameView().frame().rootFrame().frameID(), ScrollingNodeType::PluginScrolling, pluginScrollingNodeID, pluginHostingNodeID, 0);
```
Each of these is doing that rootFrame() tree walk you added; that's going to get slow.

nmoucht replied:
Alex is working on a patch to make this better.

nmoucht replied:
Fixed after https://github.com/WebKit/WebKit/pull/27823.

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -423,9 +427,11 @@ bool ScrollingTree::commitTreeStateInternal(std::unique_ptr<ScrollingStateTree>&
+        removeNode(nodeID, commitState.frameHostingNode.get());
     }
 
+    LOG_WITH_STREAM(ScrollingTree, stream << "committed ScrollingTree" << scrollingTreeAsText(debugScrollingStateTreeAsTextBehaviors));
```
Is there a reason this log was moved?

nmoucht replied:
I added this before the recursive commits below as it made debugging confusing when the scrolling trees would be printed out of order.

achristensen07 commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -109,15 +108,54 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
+    if (it != m_scrollingStateTrees.end())
+        return it->value.get();
+
+    ASSERT_WITH_MESSAGE(!page() || page()->settings().siteIsolationEnabled() || m_scrollingStateTrees.isEmpty(), "With site isolation off, we should only ever have one scoling state tree. Get the frame identifier from LocalFrame::rootFrame.");
```
This assertion can now be replaced by ASSERT(Frame::isRootFrameIdentifier(identifier)) at the beginning of the function.

nmoucht replied:
Fixed.

smfr commented on LayoutTests/http/tests/site-isolation/scrolling/remove-root-frame.html:
```diff
@@ -0,0 +1,30 @@
+    setTimeout(function() {
+        if (window.testRunner)
+            testRunner.notifyDone();
+    }, 500);
```
Why the 500ms timeout?

achristensen07 replied:
This gives time for things to happen to make the call to rootFrameWasRemoved more likely to be hit by this test.  The results are the same with or without it being hit, so it's not flaky, but this additional timeout makes it so that more code is at least hit by the test in the common case.

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.cpp:
```diff
@@ -83,6 +83,14 @@ void ScrollingTreeFrameHostingNode::willBeDestroyed()
+{
+    if (node) {
+        m_hostedChildren.remove(node);
+        m_children.removeFirst(Ref<ScrollingTreeNode>(*node));
```
The way to get a `Ref<>` from a `RefPtr<>` is to use `releaseNonNull()`

nmoucht replied:
Fixed.

nmoucht committed:
[Site Isolation] Properly create scrolling tree for web process with multiple root frames
https://bugs.webkit.org/show_bug.cgi?id=272721
rdar://126001790

Reviewed by Simon Fraser.

Adjust AsyncScrollingCoordinator to hold a scrolling state tree per root frame.
Also thanks to Alex who helped take this pr over the finishing line.

* LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree.html:
* LayoutTests/http/tests/site-isolation/scrolling/multiple-root-frames-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/multiple-root-frames.html: Copied from LayoutTests/http/tests/site-isolation/scrolling/basic-scrolling-tree.html.
* LayoutTests/http/tests/site-isolation/scrolling/remove-root-frame-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/scrolling/remove-root-frame.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/SmallMap.h: Added.
(WTF::SmallMap::ensure):
(WTF::SmallMap::remove):
(WTF::SmallMap::get const):
(WTF::SmallMap::forEach const):
(WTF::SmallMap::size const):
(WTF::SmallMap::rawStorage const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setBackForwardCacheState):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::~LocalFrame):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::rootFrameID const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::AsyncScrollingCoordinator):
(WebCore::AsyncScrollingCoordinator::stateNodeForNodeID const):
(WebCore::AsyncScrollingCoordinator::stateNodeForScrollableArea const):
(WebCore::AsyncScrollingCoordinator::ensureScrollingStateTreeForRootFrameID):
(WebCore::AsyncScrollingCoordinator::existingScrollingStateTreeForRootFrameID const):
(WebCore::AsyncScrollingCoordinator::rootFrameWasRemoved):
(WebCore::AsyncScrollingCoordinator::stateTreeForNodeID const):
(WebCore::AsyncScrollingCoordinator::willCommitTree):
(WebCore::AsyncScrollingCoordinator::updateEventTrackingRegions):
(WebCore::AsyncScrollingCoordinator::frameViewLayoutUpdated):
(WebCore::AsyncScrollingCoordinator::frameViewEventTrackingRegionsChanged):
(WebCore::AsyncScrollingCoordinator::frameViewForScrollingNode const):
(WebCore::AsyncScrollingCoordinator::createNode):
(WebCore::AsyncScrollingCoordinator::insertNode):
(WebCore::AsyncScrollingCoordinator::unparentNode):
(WebCore::AsyncScrollingCoordinator::unparentChildrenAndDestroyNode):
(WebCore::AsyncScrollingCoordinator::detachAndDestroySubtree):
(WebCore::AsyncScrollingCoordinator::clearAllNodes):
(WebCore::AsyncScrollingCoordinator::reconcileViewportConstrainedLayerPositions):
(WebCore::AsyncScrollingCoordinator::ensureRootStateNodeForFrameView):
(WebCore::AsyncScrollingCoordinator::hasSubscrollers const):
(WebCore::AsyncScrollingCoordinator::scrollingStateTreeAsText const):
(WebCore::AsyncScrollingCoordinator::commitTreeStateForRootFrameID):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
(WebCore::AsyncScrollingCoordinator::scrollingStateTrees const):
(WebCore::AsyncScrollingCoordinator::scrollingStateTree): Deleted.
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::rootFrameWasRemoved):
(WebCore::ScrollingCoordinator::createNode):
(WebCore::ScrollingCoordinator::insertNode):
(WebCore::ScrollingCoordinator::clearAllNodes):
(WebCore::ScrollingCoordinator::hasSubscrollers const):
(WebCore::ScrollingCoordinator::willCommitTree):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::removeNode):
(WebCore::ScrollingTree::commitTreeStateInternal):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.cpp:
(WebCore::ScrollingTreeFrameHostingNode::removeHostedChild):
* Source/WebCore/page/scrolling/ScrollingTreeFrameHostingNode.h:
* Source/WebCore/page/scrolling/ThreadedScrollingCoordinator.cpp:
(WebCore::ThreadedScrollingCoordinator::commitTreeStateIfNeeded):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::rootFrameID const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::maintainsEventRegion const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingLayers):
(WebCore::RenderLayerCompositor::attachWidgetContentLayersIfNecessary):
(WebCore::RenderLayerCompositor::registerScrollingNodeID):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::rootFrameID const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::rootFrameID const):
* Source/WebCore/rendering/RenderListBox.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::rootFrameID const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createScrollingNodeIfNecessary):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::buildTransaction):
* Tools/TestWebKitAPI/Tests/WTF/SmallSet.cpp:
(TestWebKitAPI::TEST(WTF_SmallMap, Basic)):

Canonical link: https://commits.webkit.org/278591@main
